README
DBIx-Log4perl version 0.07
==========================

DBIx-Log4perl is an extension for DBI to selectively log SQL, parameters,
result-sets, transactions etc to a Log::Log4perl handle.

You will need the DBI and Log::Log4perl modules first.

Please see the NOTES section in the pod for incompatibilities.

Please make sure you have an up to date Test::More before running the
tests. You will need at least 0.62.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

NOTE: To to a full test you need a working DBI and DBD with access to
a database where you can create a table. Once you have this set up
you should run perl Makefile.PL like this:

perl Makefile.PL --dsn=dsn --user=dbuser --password=dbpassword --table=table

where:

  dsn is a valid DBI connection dsn e.g. dbd::mysql::database
  dbuser is a valid user to access database
  dbpassword is the password for dbuser
    (may be omitted if not password required)
  table is the name of a table to create for testing
    (NOTE: this table will be dropped if it already exists)

DEPENDENCIES

This module requires these other modules and libraries:

  DBI 1.50
  Log::Log4perl 1.04
  Test::More 0.62

COPYRIGHT AND LICENCE

Copyright (C) 2006 by Martin J. Evans

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension DBIx::Log4perl.

0.07  Tues July 25 2006

  Fix bug in execute_array which did not handle scalars in parameter
  list for execute array. e.g.

  execute_array(undef,
	        \@p1_list, # OK
	        $p2 # was not handled properly
  )

  Added support to execute_array to automatically add an
  ArrayTupleStatus if one was not specified so DBIx::Log4perl can log
  errors in execute_array.

  Fixed bug in execute_array where comparison with logging mask
  was && instead of &.

  Fixed bug in execute_array which did not handle passing undef
  attributes as first argument to execute_array.

  Added missing support for finish method in DBI.

  Added missing support for prepare_cached method in DBI.

  Added missing support for get_info method in DBI.

0.06  Fri Jun 23 2006

  Added example captured error and description to POD.

  Fix problem where attribute DBIx_l4p_logmask was not documented
  properly and was not acted upon.

  execute_array was logging errors even when not asked to.

  Fix problem where error handler was not put in place unless some
  attributes are passed in the connect method.

  Fix unitialised string in ne in the do method if the do method in
  the driver failed.

  Changes to Captured error output:
    reformatted to be more readable
    differentiate DBI::lasth Statement from SQL on other handles
    Collect sub statements off and dbh and output any values and
      ParamValues at the end.
    Fix bug dereferencing Statement from private__DBIx_Log4perl handle
      instead of the dbh.
    Add Kids and ActiveKids output
    ParamValues for some methods where the statement was already
      destroyed e.g. a failed do, were not output in the captured
      error because they were not saved.

  Change error handler to display true number of sub statements
  (ChildHandles) under a dbh instead of the index of the last child
  handle i.e. show 0 not -1 for none. Add missing newline in output.

  Avoid problem in global cleanup with
  "(in cleanup) Can't call method "FETCH" on an undefined value" message
  when disconnect not called.

  Avoid logging of do method calls showing prepare/execute and affected
  logging.

0.05  Fri Jun 16 2006

  Fixed a few typos in the pod.

  Fixed bug in DBIx::Log4perl::prepare which was causing a non-Null
  statement handle to be returned when the driver's prepare method
  failed. When this handle was then used to execute a statement
  DBI would spot it was not a proper handle and complain with:

    dbih_getcom handle HASH(0x8c995d8) is not a DBI handle (has no magic)
    at /usr/lib/perl5/site_perl/5.8.8/DBIx/Log4perl/db.pm line 61

  and you might have seen something like the following on stderr for
  your Perl application:

    SV = RV(0x89a8e70) at 0x8e19950
    REFCNT = 1
    FLAGS = (TEMP,ROK)
    RV = 0x8c995d8

  Fixed issue in the error handler which does not handle ParamValues
  hash keys having undefined values.

  Fixed issue in _dbix_l4p_debug which was stopping calls to
  $sth->execute without any arguments being logged.

  Fixed typo in test skipped message in simple.t

0.04  Sun May 21 2006

  Additional tests mentioned in 0.03 omitted from MANIFEST

  Add requirement on Test::more 0.62 because we are using
  Test::Mores' BAIL_OUT.

  Updated README to include instructions on defining your database
  connection.

0.03  Sat May 20 2006

  If a statement is executed with $sth->execute and it changes rows
  then log the affected rows.

  Added more tests and support for setting DB login details.

0.02  Mon Apr 24 2006

  Internal error handler changes:
    The message passed to the internal error handler was never written to
      the log file.
    The $dbi Statement attribute was not being logged to the log file

  Add new dbix_l4p_logdie method

  README changes

  Don't output "execute: undef" is execute is called with no args,
  just output "execute:"

  Add $DBI::lasth->{Type} and $DBI::lasth->{Statement} output to
  error_handler

  When do called, save SQL in Statement. This is because if an app
  calls do for an insert/update/delete and the operation succeeds (in
  DBI terms) BUT the app knows it has logically failed (e.g. inserted
  0 rows when it expected 1) the app may call dbix_l4p_logdie and we
  want the SQL to be visible.

0.01  Mon Apr  3 16:10:52 2006
	- original version; created by h2xs 1.23 with options
		-A -X --compat-version=5.8.4 --use-new-tests -v 0.01 -n DBIx::Log4perl



t/lib.pl
# $Id: lib.pl 208 2006-06-21 15:10:33Z martin $
use Cwd;
use File::Spec;

my $logtmp1;
my $logtmp2;

sub get_config
{
    my @v;
    if (-f ($file = "t/dbixl4p.config")  ||
	-f ($file = "../dbixl4p.config") ||
	-f ($file = "dbixl4p.config")) {
	open IN, "$file";
	while (<IN>) {
	    chomp;
	    if ($_ eq 'UNDEF') {
		push @v, undef;
	    } else {
		push @v, $_;
	    }
	}
    }
    return @v;
}
sub config
{
    open OUT, ">pipe.pl" or die "Failed to create pipe.pl - $!";
    print OUT "#!$^X\n";
    print OUT 'while (<STDIN>) {open OUT, ">>dbixl4p.log"; print OUT $_;close OUT;}';
    close OUT;
    chmod 0777, "pipe.pl";

    $logtmp1 = File::Spec->catfile(File::Spec->tmpdir, 'dbixroot.log');
    $logtmp2 = File::Spec->catfile(File::Spec->tmpdir, 'dbix.log');
    my $cwd = getcwd();
    my $pipe = File::Spec->catfile($cwd, "pipe.pl");
    my $loginit = qq(
log4perl.logger = FATAL, LOGFILE
   
# LOGFILE appender used by root (here)
# log anything at level ERROR and above to $logtmp1
log4perl.appender.LOGFILE=Log::Log4perl::Appender::File
log4perl.appender.LOGFILE.filename=$pipe
log4perl.appender.LOGFILE.mode=pipe
log4perl.appender.LOGFILE.utf8=1
log4perl.appender.LOGFILE.autoflush=1
log4perl.appender.LOGFILE.Threshold = ERROR
log4perl.appender.LOGFILE.layout=PatternLayout
log4perl.appender.LOGFILE.layout.ConversionPattern=[%r] %F %L %c - %m%n

# Define DBIx::Log4perl to output DEBUG and above msgs to
# $logtmp2 using the simple layout
log4perl.logger.DBIx.Log4perl=DEBUG, A1
log4perl.appender.A1=Log::Log4perl::Appender::File
log4perl.appender.A1.filename=$pipe
log4perl.appender.A1.mode=pipe
log4perl.appender.A1.utf8=1
log4perl.appender.A1.autoflush=1
log4perl.appender.A1.layout=Log::Log4perl::Layout::SimpleLayout
		     );

    ok (Log::Log4perl->init(\$loginit), 'init config');

    ok (Log::Log4perl->get_logger('DBIx::Log4perl'), 'get log handle');

}

sub check_log
{
    my $s = shift;
    $$s = "";
    return 0 if (! -r "dbixl4p.log");
    my ($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size, $atime,$mtime,$ctime,$blksize,$blocks) = stat("dbixl4p.log");
    return 0 if ($size <= 0);
    open IN, "<dbixl4p.log";
    while (<IN>) {$$s .= $_};
    close IN;
    unlink "dbixl4p.log";
    diag($$s) if ($ENV{TEST_VERBOSE});
    return 1;
}
1;


lib/DBIx/Log4perl/db.pm
# $Id: db.pm 215 2006-06-28 13:58:01Z martin $
use strict;
use warnings;
use DBI;
use Log::Log4perl;

package DBIx::Log4perl::db;
@DBIx::Log4perl::db::ISA = qw(DBI::db DBIx::Log4perl);
use DBIx::Log4perl::Constants qw (:masks $LogMask);

sub prepare {
  my($dbh, @args) = @_;

  my $h = $dbh->{private_DBIx_Log4perl};
  $dbh->_dbix_l4p_debug('prepare', $args[0])
    if (($LogMask & DBIX_L4P_LOG_INPUT) &&
	  (caller !~ /^DBIx::Log4perl/) &&
          (caller !~ /^DBD::/)); # e.g. from selectall_arrayref
  my $sth = $dbh->SUPER::prepare(@args);
  $sth->{private_DBIx_Log4perl} = $h if ($sth);
  return $sth;
}

sub do {
  my ($dbh, @args) = @_;
  my $h = $dbh->{private_DBIx_Log4perl};

  $h->{Statement} = $args[0];
  $dbh->_dbix_l4p_debug('do', @args)
    if ($LogMask & DBIX_L4P_LOG_INPUT);

  my $affected = $dbh->SUPER::do(@args);

  if (!defined($affected)) {
      $dbh->_dbix_l4p_error('do error for ', @args)
	  if (($LogMask & DBIX_L4P_LOG_ERRCAPTURE) &&
	      !($LogMask & DBIX_L4P_LOG_INPUT)); # not already logged
  } elsif (defined($affected) && $affected eq '0E0' &&
	  ($LogMask & DBIX_L4P_LOG_WARNINGS)) {
      $dbh->_dbix_l4p_warning('no effect from ', @args);
  } elsif (($affected ne '0E0') && ($LogMask & DBIX_L4P_LOG_INPUT)) {
      $dbh->_dbix_l4p_debug('affected', $affected);
      $h->{logger}->debug("\t" . $dbh->SUPER::errstr)
	if (!defined($affected));
  }
  return $affected;
}

sub selectrow_array {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    $dbh->_dbix_l4p_debug('selectrow_array', @args)
      if ($LogMask & DBIX_L4P_LOG_INPUT);

    if (wantarray) {
	my @ret = $dbh->SUPER::selectrow_array(@args);
	$dbh->_dbix_l4p_debug('result', @ret)
	  if ($LogMask & DBIX_L4P_LOG_OUTPUT);
	return @ret;

    } else {
	my $ret = $dbh->SUPER::selectrow_array(@args);
	$dbh->_dbix_l4p_debug('result', $ret)
	  if ($LogMask & DBIX_L4P_LOG_OUTPUT);
	return $ret;
    }
}

sub selectrow_arrayref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    $dbh->_dbix_l4p_debug('selectrow_arrayref', @args)
      if ($LogMask & DBIX_L4P_LOG_INPUT);

    my $ref = $dbh->SUPER::selectrow_arrayref(@args);
    $dbh->_dbix_l4p_debug('result', $ref)
      if ($LogMask & DBIX_L4P_LOG_OUTPUT);
    return $ref;
}

sub selectrow_hashref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    $dbh->_dbix_l4p_debug('selectrow_hashref', @args)
      if ($LogMask & DBIX_L4P_LOG_INPUT);

    my $ref = $dbh->SUPER::selectrow_hashref(@args);
    # no need to show result - fetch will do this
    return $ref;

}

sub selectall_arrayref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    $dbh->_dbix_l4p_debug('selectall_arrayref', @args)
      if ($LogMask & DBIX_L4P_LOG_INPUT);

    my $ref = $dbh->SUPER::selectall_arrayref(@args);
    $dbh->_dbix_l4p_debug('result', $ref)
      if ($LogMask & DBIX_L4P_LOG_OUTPUT);
    return $ref;
}

sub selectall_hashref {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    $dbh->_dbix_l4p_debug('selectall_hashref', @args)
      if ($LogMask & DBIX_L4P_LOG_INPUT);

    my $ref = $dbh->SUPER::selectall_hashref(@args);
    # no need to show result - fetch will do this
    return $ref;

}
sub disconnect {
    my $dbh = shift;

    if ($dbh) {
	my $h;
	eval {
	    # Avoid 
	    # (in cleanup) Can't call method "FETCH" on an undefined value
	    $h = $dbh->{private_DBIx_Log4perl};
	};
	if (!$@) {
	    $h->{logger}->debug("disconnect")
		if ($LogMask & DBIX_L4P_LOG_CONNECT);
	}
    }
    return $dbh->SUPER::disconnect;
    
}

sub begin_work {
    my $dbh = shift;
    my $h = $dbh->{private_DBIx_Log4perl};
    $h->{logger}->debug("start transaction")
      if ($LogMask & DBIX_L4P_LOG_TXN);

    return $dbh->SUPER::begin_work;
}

sub rollback {
    my $dbh = shift;
    my $h = $dbh->{private_DBIx_Log4perl};
    $h->{logger}->debug("roll back")
      if ($LogMask & DBIX_L4P_LOG_TXN);

    return $dbh->SUPER::rollback;
}
  
sub commit {
    my $dbh = shift;

    my $h = $dbh->{private_DBIx_Log4perl};
    $h->{logger}->debug("commit")
      if ($LogMask & DBIX_L4P_LOG_TXN);

    return $dbh->SUPER::commit;
}

sub last_insert_id {
    my ($dbh, @args) = @_;

    my $h = $dbh->{private_DBIx_Log4perl};
    $h->{logger}->debug(
	sub {Data::Dumper->Dump([\@args], ['last_insert_id'])})
      if ($LogMask & DBIX_L4P_LOG_INPUT);

    my $ret = $dbh->SUPER::last_insert_id(@args);
    $h->{logger}->debug(sub {"\t" . DBI::neat($ret)})
      if ($LogMask & DBIX_L4P_LOG_INPUT);
    return $ret;
}
1;


lib/DBIx/Log4perl/st.pm
# $Id: st.pm 265 2006-08-04 15:51:00Z martin $
use strict;
use warnings;
use DBI;
use Log::Log4perl;

package DBIx::Log4perl::st;
@DBIx::Log4perl::st::ISA = qw(DBI::st DBIx::Log4perl);
use DBIx::Log4perl::Constants qw (:masks $LogMask);

sub finish {
    my ($sth) = shift;
    my $h = $sth->{private_DBIx_Log4perl};
    
    $sth->_dbix_l4p_debug('finish')
	if ($h->{logmask} & DBIX_L4P_LOG_INPUT);
    return $sth->SUPER::finish;
}
	    
sub execute {
    my ($sth, @args) = @_;
    my $h = $sth->{private_DBIx_Log4perl};

    $sth->_dbix_l4p_debug('execute', @args)
      if (($h->{logmask} & DBIX_L4P_LOG_INPUT) && (caller !~ /^DBD::/));

    my $ret = $sth->SUPER::execute(@args);

    if (!$ret) {		# error
	$h->{logger}->error("\tfailed with " . $sth->errstr)
	    if (($h->{logmask} && DBIX_L4P_LOG_ERRCAPTURE) && # logging errors
		(caller !~ /^DBD::/)); # not called from DBD e.g. execute_array
    } elsif (defined($ret)) {
        $sth->_dbix_l4p_debug('affected', $ret)
	    if ((!defined($sth->{NUM_OF_FIELDS})) && # not a result-set
		($h->{logmask} & DBIX_L4P_LOG_INPUT)	&& # logging input
		(caller !~ /^DBD::/));
    }
    return $ret;
}

sub execute_array {
    my ($sth, @args) = @_;
    my $h = $sth->{private_DBIx_Log4perl};

    $sth->_dbix_l4p_debug('execute_array', @args)
      if ($h->{logmask} & DBIX_L4P_LOG_INPUT);

    if (($#args >= 0) && ($args[0]) &&
	    (ref($args[0]) eq 'HASH') &&
		(!exists($args[0]->{ArrayTupleStatus}))) {
	$args[0]->{ArrayTupleStatus} = \my @tuple_status;
    } elsif (!$args[0]) {
	$args[0] = {ArrayTupleStatus => \my @tuple_status};
    }
    my $executed = $sth->SUPER::execute_array(@args);
    if (!$executed) {
        #print Data::Dumper->Dump([$sth->{ParamArrays}], ['ParamArrays']), "\n";
	return $executed unless ($h->{logmask} & DBIX_L4P_LOG_ERRORS);
	return $executed if (!$args[0] || ref($args[0] ne 'HASH') ||
			       !exists($args[0]->{ArrayTupleStatus}));
        my $pa = $sth->{ParamArrays};
        if (scalar(@args) > 0) {
	    $h->{logger}->error("execute_array error:");
	  my $ats = $args[0]->{ArrayTupleStatus};
	  for my $n (0..@{$ats}-1) {
	    next if (!ref($ats->[$n]));
	    $sth->_dbix_l4p_error('Error', $ats->[$n]);
	    my @plist;
	    foreach my $p (keys %{$pa}) {
		if (ref($pa->{$p})) {
		    push @plist, $pa->{$p}->[$n];
		} else {
		    push @plist, $pa->{$p};
		}
	    }
	    $h->{logger}->error(sub {"\t for " . join(',', @plist)});
	  }
	}
    }
    return $executed;
}

sub bind_param {
  my($sth, @args) = @_;
  my $h = $sth->{private_DBIx_Log4perl};

  $sth->_dbix_l4p_debug('bind_param', @args)
    if ($h->{logmask} & DBIX_L4P_LOG_INPUT);
  return $sth->SUPER::bind_param(@args);
}

sub bind_param_inout {
  my($sth, @args) = @_;
  my $h = $sth->{private_DBIx_Log4perl};

  $sth->_dbix_l4p_debug('bind_param_inout', @args)
    if ($h->{logmask} & DBIX_L4P_LOG_INPUT);
  return $sth->SUPER::bind_param_inout(@args);
}

sub bind_param_array {
  my($sth, @args) = @_;
  my $h = $sth->{private_DBIx_Log4perl};

  $sth->_dbix_l4p_debug('bind_param_array', @args)
    if ($h->{logmask} & DBIX_L4P_LOG_INPUT);
  return $sth->SUPER::bind_param_array(@args);
}

sub fetch {			# alias for fetchrow_arrayref
  my($sth, @args) = @_;
  my $h = $sth->{private_DBIx_Log4perl};

  my $res = $sth->SUPER::fetch(@args);
  $h->{logger}->debug(sub {Data::Dumper->Dump([$res], ['fetch'])})
    if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
  return $res;
}

sub fetchrow_arrayref {			# alias for fetchrow_arrayref
  my($sth, @args) = @_;
  my $h = $sth->{private_DBIx_Log4perl};

  my $res = $sth->SUPER::fetchrow_arrayref(@args);
  $h->{logger}->debug(sub {Data::Dumper->Dump([$res],
						   ['fetchrow_arrayref'])})
    if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
  return $res;
}

sub fetchrow_array {
  my ($sth, @args) = @_;
  my $h = $sth->{private_DBIx_Log4perl};

  my @row = $sth->SUPER::fetchrow_array(@args);
  $h->{logger}->debug(sub {
			     Data::Dumper->Dump([\@row], ['fetchrow_array'])})
    if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
  return @row;
}

sub fetchrow_hashref {
  my($sth, @args) = @_;
  my $h = $sth->{private_DBIx_Log4perl};

  my $res = $sth->SUPER::fetchrow_hashref(@args);
  $h->{logger}->debug(
      sub {Data::Dumper->Dump([$res], ['fetchrow_hashref'])})
    if ($h->{logmask} & DBIX_L4P_LOG_OUTPUT);
  return $res;
}

1;


lib/DBIx/Log4perl/Constants.pm
use strict;
use warnings;

package DBIx::Log4perl::Constants;
require Exporter;
our @ISA = qw(Exporter);

use constant DBIX_L4P_LOG_DEFAULT => 0x1d;
use constant DBIX_L4P_LOG_ALL => 0xffffff;
use constant DBIX_L4P_LOG_INPUT => 0x1;
use constant DBIX_L4P_LOG_OUTPUT => 0x2;
use constant DBIX_L4P_LOG_CONNECT => 0x4;
use constant DBIX_L4P_LOG_TXN => 0x8;
use constant DBIX_L4P_LOG_ERRCAPTURE => 0x10;
use constant DBIX_L4P_LOG_WARNINGS => 0x20;

our $LogMask = DBIX_L4P_LOG_DEFAULT;

our @EXPORT = ();
our @EXPORT_OK = qw ($LogMask);
our @EXPORT_MASKS = qw(DBIX_L4P_LOG_DEFAULT
		       DBIX_L4P_LOG_ALL
		       DBIX_L4P_LOG_INPUT
		       DBIX_L4P_LOG_OUTPUT
		       DBIX_L4P_LOG_CONNECT
		       DBIX_L4P_LOG_TXN
		       DBIX_L4P_LOG_ERRCAPTURE
		       DBIX_L4P_LOG_WARNINGS
		     );
our %EXPORT_TAGS = (masks => \@EXPORT_MASKS);
Exporter::export_ok_tags('masks');


1;


Makefile.PL
use 5.008000;
use ExtUtils::MakeMaker;
use Getopt::Long();
use Data::Dumper;

use vars qw($opt);
$opt = {};

Getopt::Long::GetOptions(
    $opt,
    "help" => sub {Usage()},
    "dsn=s",
    "user=s",
    "password:s",
    "table=s",
    ) || die Usage();


my ($dsn, $user, $password, $table) = ('', '', 'UNDEF', '');

$dsn = $ENV{DBI_DSN} if (exists($ENV{DBI_DSN}));
$user = $ENV{DBI_USER} if (exists($ENV{DBI_USER}));
$password = $NV{DBI_PASS} if (exists($ENV{DBI_PASS}));
$table = $ENV{DBI_TABLE} if (exists($ENV{DBI_TABLE}));

$dsn = $opt->{dsn} if (exists($opt->{dsn}));
$user = $opt->{user} if (exists($opt->{user}));
$password = $opt->{password} if (exists($opt->{password}));
$table = $opt->{table} if (exists($opt->{table}));

open OUT, ">dbixl4p.config";
print OUT "$dsn\n";
print OUT "$user\n";
print OUT "$password\n";
print OUT "$table\n";
close OUT;

warn "\nPlease NOTE\nYou need to set and export the DBI_* environment variables or supply suitable values on the command line (see Perl Makefile.PL help) to run a complete test\n\n";

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'DBIx::Log4perl',
    VERSION_FROM      => 'lib/DBIx/Log4perl.pm', # finds $VERSION
    PREREQ_PM         => {Log::Log4perl => 1.04,
		          DBI => 1.50,
		          Test::More => 0.62},
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/DBIx/Log4perl.pm', # retrieve abstract from module
       AUTHOR         => 'Martin J. Evans <mjevans@cpan.org>') : ()),
);


sub Usage {
  print STDERR <<"USAGE";
Usage: perl $0 [options]

Possible options are:

  --dsn=<db>         The DBI DSN to pass to DBI->connect
  --user=<user>      Use the username <user> for running the test suite;
                     defaults to no username
  --password=<pwd>   Use the password <pwd> for running the test suite;
                     defaults to no password
  --table=<table>    A table to be used as a test table in the DB
                     defaults to dbixl4p
USAGE
  exit 1;
}
